### PR TITLE
MM-37165: Fix improper plugin shutdown

### PR DESCRIPTION
**Pull request by agnivade**
This was caught from a race test failure. While the failure manifested due to a log being
written from a test after the test exited, the real reason was hidden further deeper.

What was happening is that the server would always listen for plugin requests in a
separate goroutine via `g.muxBroker.AcceptAndServe` in the `OnActivate` hook. But the
plugin shutdown process would just close the plugin connections and move on, leading
to the classic case of improperly shut down goroutines.

When this happened, an opportunity opens up in a way that the server
would still be executing a request whereas the main goroutine and therefore the parent
test has already finished. This would lead to an error like

```
{"level":"error","ts":1626451258.4141135,"caller":"mlog/sugar.go:25","msg":"pluginAPI scheduleOnce poller encountered an error but is still polling","plugin_id":"com.mattermost.plugin-incident-management","error":"ListPluginKeys: Unable to list all the plugin keys., failed to get PluginKeyValues with pluginId=com.mattermost.plugin-incident-management: sql: database is closed
```

And now, this finally calls `mlog.Error`, which finally triggers our race condition :)

To fix this, we use some basic synchronization via waitgroups and just wait for it
to finish after closing the plugin process.

https://mattermost.atlassian.net/browse/MM-37165

```release-note
NONE
```

### CODE REVIEW

**Review by lieut-data — APPROVED** [review]

File: plugin/client_rpc.go [hunk 1]
```diff
@@ -240,15 +242,22 @@ type Z_OnActivateReturns struct {
 
 func (g *hooksRPCClient) OnActivate() error {
 	muxId := g.muxBroker.NextId()
-	go g.muxBroker.AcceptAndServe(muxId, &apiRPCServer{
-		impl:      g.apiImpl,
-		muxBroker: g.muxBroker,
-	})
+	g.doneWg.Add(2)
```
[the remark this answers is not in the document]
Comment (agnivade) [reply plugin/client_rpc.go:245]: Haha, I knew you would say this. Sure, will update. :+1:

File: plugin/supervisor.go [hunk 2]
```diff
@@ -104,6 +110,11 @@ func (sup *supervisor) Shutdown() {
 	if sup.client != nil {
 		sup.client.Kill()
 	}
+
+	// Wait for API RPC server and DB RPC server to exit.
+	if sup.hooksClient != nil {
+		sup.hooksClient.doneWg.Wait()
```
[the remark this answers is not in the document]
Comment (agnivade) [reply plugin/supervisor.go:116]: I would lean on @lieut-data on that. But I think it's unlikely as those are only called from the server and the server either starts/stops a plugin, but it doesn't do that concurrently.

**Review by streamer45 — APPROVED** [review]

**Review by crspeller — APPROVED** [review]

**Comment by agnivade** [comment]
@DHaussermann - Would you be able to test couple of plugins and start/stop/remove them from the console, and verify everything works? Thanks.

**Review by DHaussermann — APPROVED** [review]
tested and passed
- Tested install and removal of various plugins
- Tested enable disable of plugins
- Tested on single and HA environment (including with aws-multitenant-rds-postgres DB)
- No unexpected errors found 

LGTM!

**Pull request merged**